Add documentation (#87)
precice/systemtests: docs/main.md

> # General architecture 

Reviewer: Naming this file `README.md` would make it shown by default on GitHub when one is at the `docs/` directory. Currently it needs additional thought to select the right one and it will become more difficult as the project grows.

> 

Reviewer: Maybe add a few lines to describe what this file is and what other files there are under `docs/`?

Author: `README` is a good idea. Though I would not move `main.md` into `README`, rather I would use it to describe different documentation files, as you suggest here.

Reviewer: That's also good! :+1:

Tests are executed in several stages: 

## Building preCICE

> Several images of preCICE are built, based on different system (Ubuntu 16.04/18.04, Arch Linux) and different specifications

Reviewer: ```suggestion
Several images of preCICE are built, based on different systems (Ubuntu 16.04/18.04, Arch Linux) and different specifications
```

> of properties of the image ( root/non-root, package, different MPI versions ). This corresponds to the `Dockerfile.$systemname$.${system_spec}`

Reviewer: Where are these properties defined? I know what `package` means in this context, but I am not sure if a new developer would immediately understand it.

> Resultant images are then pushed to the dockerhub.

Reviewer: A link would be useful:
```suggestion
Resultant images are then pushed to [Dockerhub](https://hub.docker.com/u/precice).
```

## Building solvers 

> Is done once for every solver, using `Dockefile.${solver_name}`. Resultant images are then stored on the dockerhub.

Reviewer: A link would be useful:
```suggestion
Is done once for every solver, using `Dockefile.${solver_name}`. Resultant images are then pushed to [Dockerhub](https://hub.docker.com/u/precice).
```

## Building adapters

> To build the adapter we follow two-stage builds, first importing from the preCICE image then from the solvers image (if the solvers image is

Reviewer: I think it reads better like this:
```suggestion
To build an adapter we follow a two-stage build, first importing from the preCICE image and then importing from the solver's image (if the solver's image is
```

actually relevant for building the adapter). We can import from different preCICE base images by specifying `from` argument during building of the

> the image, e.g:

Reviewer: Duplicate
```suggestion
image, e.g:
```

```

> docker build --build-arg from=precice/precice-ubuntu1804.home-develop -f Dockerfile.openfoam-adapter -t openfoam-adapter-ubuntu1804.home-develop .,

Reviewer: Typo?
```suggestion
docker build --build-arg from=precice/precice-ubuntu1804.home-develop -f Dockerfile.openfoam-adapter -t openfoam-adapter-ubuntu1804.home-develop .
```

> docker build --build-arg from=precice/precice-ubuntu1804.home-develop -f Dockerfile.openfoam-adapter -t openfoam-adapter-ubuntu1804.home-develop .,

Reviewer: I tried this, it worked out-of-the box! :smile:

```

> Note, that due to the fact that different adapter 'adapts' differently, the general procedure is slightly

Reviewer: ```suggestion
Note, that due to the fact that each adapter "adapts" differently, the general procedure is slightly
```

different and is outlined below: 

- **SU2 adapter**

> 

Reviewer: For consistency with the rest:
```suggestion

```

>   Since SU2 adapter modifies SU2 source code, it makes no sense to copy clean SU2 build, because it will need to be rebuilt. Instead we just build the adapter

Reviewer: ```suggestion
  Since the SU2 adapter modifies the SU2 source code, it makes no sense to copy a clean SU2 build, because it will need to be rebuilt. Instead, we directly build the adapted
```

>   from  SU2 source after applying the necessary modifications.

Reviewer: Together with the previous suggestion
```suggestion
  SU2 source.
```

> - **dealii adapter**

Reviewer: ```suggestion
- **deal.II adapter**
```

>   Since dealii is a library and the adapter just needs to link with it, here we actually perform the import of dealii image and only build

Reviewer: ```suggestion
  Since deal.II is a library and the adapter just needs to link with it, here we import the previously-built deal.II image and only build
```

>   the actual adapter

Reviewer: ```suggestion
  the adapter
```

- **CalculiX adapter**

>   We only import the necessary libraries (Spooles, ARPACK), since CalculiX adapter modifies CalculiX source.

Reviewer: ```suggestion
  We only import the necessary libraries (Spooles, ARPACK), since the CalculiX adapter modifies the CalculiX source code.
```

> - **openFOAM adapter**

Reviewer: ```suggestion
- **OpenFOAM adapter**
```

>   Same as for deal.ii adapter.

Reviewer: ```suggestion
  Same as for the deal.II adapter.
```

- **Nutils adapter**

>   There is no Nutils adapter. There is python script that communicates with Nutils and preCICE. We therefore provide Dockefile for it just for the where it is used,

Reviewer: ```suggestion
  The Nutils adapter is a Python script that communicates with Nutils and preCICE. We therefore only provide a Nutils adapter Dockefile for the test running stage (not for the building stage).
```

>   and don't build during adapter building stage, but during test running stage.

Reviewer: Together with the previous suggestion:
```suggestion
```

- **FEniCS adapter**

>   We install FEniCS, together with adapter using `Dockefile.fenics-adapter` file.

Reviewer: ```suggestion
  We install FEniCS, together with the FEniCS adapter, using the `Dockefile.fenics-adapter` file.
```

> All adapters are built using user `precice` with `gid` and `uid` equal to 1000 and of `precice` group. As a last build stage they create folder for the volume mapping in 

Reviewer: ```suggestion
All adapters are built using user `precice` with `gid` and `uid` equal to 1000 and of `precice` group. As a last building stage they create folder for the file system volume mapping in the
```
I added the `file system` because `volume mapping` could mean a different thing in preCICE.

`/home/precice/Data` folder.

## Running tests

> At this stage, we spawn containers with the necessary solvers for the simulation. The execution is controlled using [docker-compose](https://docs.docker.com/compose/). Separate volumes are created for the input and the output of every adapter. Additional docker network is created and used for the communication between containers. Exchange of the data is done using another common volume. The general pattern across of the tests for such volumes to have locations `/home/precice/Data/Exchange`, `/home/precice/Data/Input`,

Reviewer: Minor language suggestions:
```suggestion
At this stage, we spawn containers with the necessary solvers for the simulation. The execution is controlled using [docker-compose](https://docs.docker.com/compose/). Separate volumes are created for the input and the output of every adapter. An additional docker network is created and used for the communication between containers. Exchange of data is done using an additional, common volume. The general pattern across the tests for such volumes are paths such as `/home/precice/Data/Exchange`, `/home/precice/Data/Input`,
```

`/home/precice/Data/Output`, as explained above.

> The tests are based on the corresponding tutorials from the [tutorials repository](https://github.com/precice/tutorials). Several modifications are necessary to decouple the input to solvers, adjust simulation time, or change solver version and to modify `precice-config.xml` for simulation with docker. This is done with `Dockerfile.tutorial_data` image, that clones the tutorial data, adjusts and splits the input to different solvers. It is then is used as first container spawned by `docker-compose`.

Reviewer: What do you mean by "to decouple the input to solvers"?

Author: We want the input to different adapters to be in different directories.

Reviewer: I understoon this later on, but maybe a rephrasing would help, e.g.

> The tutorials are prepared to run in the same environment and several modifications are necessary to run the solvers in different containers and testing conditions: distribute the input files, adjust the simulated time, adjust the solver version, and modify the `precice-config.xml`.

What do you think?

> The tests are based on the corresponding tutorials from the [tutorials repository](https://github.com/precice/tutorials). Several modifications are necessary to decouple the input to solvers, adjust simulation time, or change solver version and to modify `precice-config.xml` for simulation with docker. This is done with `Dockerfile.tutorial_data` image, that clones the tutorial data, adjusts and splits the input to different solvers. It is then is used as first container spawned by `docker-compose`.

Reviewer: Minor language suggestions:
```suggestion
The tests are based on the corresponding tutorials from the [tutorials repository](https://github.com/precice/tutorials). Several modifications are necessary to decouple the input to solvers, to adjust the simulated time, or to potentially change a solver's version, as well as to modify the `precice-config.xml` for the communication pattern we use in docker. This is done using the `Dockerfile.tutorial_data` image, which clones the tutorial data and adjusts and splits the input to different solvers. It is then used as the first container spawned by `docker-compose`.
```

> Afterwards, when the simulation finishes the output is copied from all the solvers into the common volume and compared to the reference output. An error is thrown if they don't match. Since it is possible for the output of the identical tests to differ in floating points due to specifics of the machine and OS, after simple comparison of the output file by file it is also compared numerically using `compare_results.sh`. It strips aways non-numerical data and then compared files field by field with respect to specified relative tolerance.

Reviewer: Minor language suggestions:
```suggestion
Afterwards, when the simulation finishes, the output is copied from all the solvers into the common volume and is compared to the reference output. An error is thrown if they don't match.

Since it is possible for the output of two identical tests to differ in floating points due to a different machine or OS (as we test different systems and properties), after simple, file by file comparison of the output, the files are also compared numerically using `compare_results.sh`. This script strips away non-numerical data and then compares files field by field with respect to a specified relative tolerance.
```

Is my suggestion `due to a different machine and OS (as we test different systems and properties)` valid?

> Afterwards, when the simulation finishes the output is copied from all the solvers into the common volume and compared to the reference output. An error is thrown if they don't match. Since it is possible for the output of the identical tests to differ in floating points due to specifics of the machine and OS, after simple comparison of the output file by file it is also compared numerically using `compare_results.sh`. It strips aways non-numerical data and then compared files field by field with respect to specified relative tolerance.

Reviewer: Which is the default tolerance? Where is this defined?

What happens when a test fails in the first comparison but succeeds in the second? Why are we confident that we don't get false successes?

Author: Default tolerance is set in the `compare_results.sh` (it can also adjusted via command line args). The general documentation of all numerical comparison is done there, so I don't duplicate it here.

> Why are we confident that we don't get false successes? 

I did some experiments on my machine with changing the results to be above the tolerance and checking if the error is reported and it is. Of course it is still hard to judge what should be the limit of the tolerance, etc, but given the main cause of failed tests are usually building of adapters of some communication problem and rarely numerics,  I would not put much focus here, until we get more failing cases due to numerics.

Reviewer: Usually the building part we already test while developing and it is not the/my main reason we rely on automated testing in this case. We really don't want to introduce numerical changes without noticing, so going conservative in this case would help.

I see that in the `compare_results.sh` we have default values:
```
avg_diff_limit="0.01"
max_diff_limit="0.01"
```
as well as values from arguments (which, if I understand correctly, are not used in `system_testing.py`). So,one needs to know that the tests will complain if the results differ more than 1% (but this default threshold is [another discussion](https://github.com/precice/systemtests/issues/90)).

The information that is missing here (and causes some confusion to an adapter developer) is which values the scripts are used with. So, maybe a clarification would help:

> The default value for the tolerance is set in `compare_results.sh` and can be changed, however the script `system_testing.py` (Travis) uses the default values.

> In either case after tests finishes, all volumes, networks and containers are removed.

Reviewer: ```suggestion
In either case, after a test finishes, all the related volumes, networks and containers are removed.
```
